timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java

```diff
@@ -34,27 +38,48 @@ public AvailableTimestamps(LastReturnedTimestamp lastReturnedTimestamp, Persiste
         this.upperLimit = upperLimit;
     }
 
-    public synchronized TimestampRange handOut(long numberToHandOut) {
-        checkArgument(
-                numberToHandOut <= MAX_TIMESTAMPS_TO_HAND_OUT,
-                "Can only hand out %s timestamps at a time, but %s were requested",
-                MAX_TIMESTAMPS_TO_HAND_OUT, numberToHandOut);
+    private static void checkValidTimestampRangeRequest(long numberToHandOut) {
+        if (numberToHandOut > MAX_TIMESTAMPS_TO_HAND_OUT) {
+            throw new IllegalArgumentException(String.format(
+                    "Can only hand out %s timestamps at a time, but %s were requested",
+                    MAX_TIMESTAMPS_TO_HAND_OUT, numberToHandOut));
+        }
+    }
 
-        long targetTimestamp = lastHandedOut() + numberToHandOut;
-        DebugLogger.logger.trace("Handing out {} timestamps, taking us to {}.", numberToHandOut, targetTimestamp);
-        return handOutTimestamp(targetTimestamp);
+    public TimestampRange handOut(long numberToHandOut) {
+        checkValidTimestampRangeRequest(numberToHandOut);
+        synchronized (this) {
+            long targetTimestamp = lastHandedOut() + numberToHandOut;
+            DebugLogger.logger.trace("Handing out {} timestamps, taking us to {}.", numberToHandOut, targetTimestamp);
+            return handOutTimestamp(targetTimestamp);
```

Comment:
May not matter relative to the other more expensive operations, but could pull the logging out of the synchronized block if we're trying to keep this as minimized as possible.